### ISSUE TITLE
Compiler crash (missing body) with declaration of inline proc

### ISSUE THREAD

**Issue by starsiderfox** [nim-works/nimskull#1368]
Traceback:

	/home/user/nimskull/compiler/nim.nim(154) nim
	/home/user/nimskull/compiler/nim.nim(89) handleCmdLine
	/home/user/nimskull/compiler/front/main.nim(539) mainCommand
	/home/user/nimskull/compiler/front/main.nim(491) compileToBackend
	/home/user/nimskull/compiler/front/main.nim(217) commandCompileToC
	/home/user/nimskull/compiler/backend/cbackend.nim(406) generateCode
	/home/user/nimskull/compiler/backend/cbackend.nim(478) generateCode
	/home/user/nimskull/compiler/backend/cbackend.nim(289) emit
	/home/user/nimskull/lib/system/assertions.nim(38) failedAssertImpl
	/home/user/nimskull/lib/system/assertions.nim(28) raiseAssert
	/home/user/nimskull/lib/system/fatal.nim(50) sysFatal
	Error: unhandled exception: /home/user/nimskull/compiler/backend/cbackend.nim(289, 10) `prc.body.code != nil` missing body [AssertionDefect]

I can only trigger this bug with 4 separate files. Minimum example is attached (try to compile `test.nim`).
[missing_body.tar.gz](https://github.com/user-attachments/files/16097122/missing_body.tar.gz)

I think the key lines are these:

	proc foo*[T](utype: typedesc[T]): var seq[T] {.inline.}
	import ./bar
	proc foo*[T](utype: typedesc[T]): var seq[T] {.inline.} = discard

Notice how the proc is declared before importing `bar` because it's part of a cyclic module import and both modules are meant to be able to use each other procs.

A workaround is easy. The biggest problem with this bug is the lack of any feedback on what the problem might be. Disallowing this kind of use of inline functions is fine, but it should have an appropriate error message.

**Comment by mrgaturus** [comment]
without `{.inline.}` it produces a linker error:
```
...........................................................
~/Programming/skulls/missing_body/foo.nim(2, 8) Warning: imported and not used: 'bar' [UnusedImport]
~/Programming/skulls/missing_body/all.nim(1, 8) Warning: imported and not used: 'foo' [UnusedImport]
/usr/bin/ld: ~/.cache/nimskull/test_d/@mbar.nim.c.o: in function `crash__bar_2':
@mbar.nim.c:(.text+0x67): undefined reference to `foo__bar_3'
/usr/bin/ld: ~/Programming/skulls/missing_body/test: hidden symbol `foo__bar_3' isn't defined
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
execution of an external program 'gcc   -o ~/Programming/skulls/missing_body/test  ~/.cache/nimskull/test_d/stdlib_ansi_c.nim.c.o ~/.cache/nimskull/test_d/stdlib_memory.nim.c.o ~/.cache/nimskull/test_d/stdlib_digitsutils.nim.c.o ~/.cache/nimskull/test_d/stdlib_assertions.nim.c.o ~/.cache/nimskull/test_d/stdlib_dollars.nim.c.o ~/.cache/nimskull/test_d/stdlib_repr_v2.nim.c.o ~/.cache/nimskull/test_d/stdlib_io.nim.c.o ~/.cache/nimskull/test_d/stdlib_system.nim.c.o ~/.cache/nimskull/test_d/@mbar.nim.c.o ~/.cache/nimskull/test_d/@mtest.nim.c.o    -ldl' failed with exit code '1'
```

**Comment by zerbina** [comment]
I think I know what the problem is. Minimized example:
```nim
# a.nim
proc forwarded*[T]()
import b
proc forwarded[T]() =
  discard

# b.nim
import a
forwarded[int]()
```

The list of instantiated generic procedures (`PContext.generics`) is per-module state, meaning that when the forward declaration is completed (in `a`), there's no access to the `forwarded[int]` instantiation, thus its body cannot be patched and `forwarded[int]` stays a bodyless procedure.

**Comment by saem** [comment]
We should probably bail on the cycle, where `b` imports `a`, but we don't have a definition for all as of yet exported routines.

**Comment by zerbina** [comment]
> We should probably bail on the cycle, where `b` imports `a`, but we don't have a definition for all as of yet exported routines.

Either that or `a` is only allowed to use already-existing instantiations:
```nim
# a.nim
proc forwarded*[T]()
# create an instantiation in the "home" module:
const _ = forwarded[int]
# now start off the cycle:
import b
# complete the forward declaration:
proc forwarded[T]() =
  discard

# b.nim
import a
forwarded[int]() # works; was already instantiated in the home module
forwarded[float]() # error: cannot instantiate foreign forward-declared generic routine
```

Personally, my preference is with only reporting an error when trying to create a new instantiation of a foreign generic routine. It's slightly easier to implement, and also fits with the current template-based nature of generic routines.

**Comment by saem** [comment]
I'm personally not a fan of circular imports whatsoever, but this issue feels rather disconnected from the better understood parts of the language, so I don't feel that I have a clear line of sight on a good solution. I think going with the one that's easier to implement for now is fine.

**Comment by zerbina** [comment]
Fixed by #1374. A proper error is now reported.
